web-ui/src/main/resources/catalog/views/default/module.js

```diff
@@ -252,50 +252,23 @@
       });
 
       $scope.resultviewFns = {
-        addMdLayerToMapSimple: function (link, md) {
-          if (gnMap.isLayerInMap(viewerMap,
-              link.name, link.url)) {
-            return;
-          }
-          var loadLayerPromise;
-
-          // handle WMTS layer info
-          if (link.protocol.indexOf('WMTS') > -1) {
-            loadLayerPromise = gnMap.addWmtsFromScratch(
-              viewerMap, link.url, link.name, undefined, md);
-          } else {
-            loadLayerPromise = gnMap.addWmsFromScratch(
-              viewerMap, link.url, link.name, undefined, md);
-          }
-
-          loadLayerPromise.then(function (layer) {
-            if (layer) {
-              gnMap.feedLayerWithRelated(layer, link.group);
-            }
-          });
-        },
-        // Add to basket first and then trigger add to map
         addMdLayerToMap: function (link, md) {
-          if (gnMap.isLayerInMap(viewerMap,
-              link.name, link.url)) {
-            return;
-          }
-          var loadLayerPromise;
-
-          // handle WMTS layer info
-          if (link.protocol.indexOf('WMTS') > -1) {
-            loadLayerPromise = gnMap.addWmtsFromScratch(
-              viewerMap, link.url, link.name, undefined, md);
-          } else {
-            loadLayerPromise = gnMap.addWmsFromScratch(
-              viewerMap, link.url, link.name, undefined, md);
+          var config = {
+            uuid: md.getUuid(),
+            type: link.protocol.indexOf('WMTS') > -1 ? 'wmts' : 'wms',
+            url: link.url
+          };
+
+          if (link.name !== '') {
+            config.name = link.name;
+            config.group = link.group;
           }
 
-          loadLayerPromise.then(function(layer) {
-            if(layer) {
-              gnMap.feedLayerWithRelated(layer, link.group);
-            }
-          });
+          // This is probably only a service
+          // Open the add service layer tab
+          $location.path('map').search({
```

Comment:
I'm not sure the "add service" panel state is relevant to appear in the URL. I think you should use `gnViewerService.openWmsTab(...)`, see:
https://github.com/geonetwork/core-geonetwork/blob/6e80de8c7e451af8df0d59ab90e61d90099b8f3d/web-ui/src/main/resources/catalog/components/viewer/ViewerService.js#L79

Reply:
But if I do that, then there is no way to open the mapviewer with a specific layers or map ? Do you have any suggestions on how to handle that ? This is a requirement in number of projects in order to be able to link third party apps/site to the map viewer. We also discussed long time ago that we need to use the location to store map states (eg. zoom, center, background ...) to be able to use the history and be able to init the mapviewer properly from URL like we do in search.

Reply:
For your first question, why not adding the layers directly to the `viewerMap` object? You should have access to it.
Using the URL to represent the map state would be nice indeed, I agree with you, but I think it has to hold a *state* and not an *action* (i.e. "add this layer"). For example URL params could be based on the way maps are configured in the UI settings:
* context file
* layers
* view extent

Reply:
> to hold a state and not an action (i.e. "add this layer")

So then what do you propose ? Use URL parameters and not angular location ones ?

> But if I do that, then there is no way to open the mapviewer with a specific layers or map ? 

@fgravin any suggestions on this ?

Reply:
Right now you can open the map with a specific context or a layer. But yes you can't compose your map from the url. I agree @jahow, if you want to have a permalink that manages this, you have to keep the state in the permalink, and not ask for actions from the url (as it's done right now for `wmsContext` and `wmsLayer` params).

I would use angular params though. You'll need to have a mecanism that saves `$location` params depending on hash (when you switch from catalog to map you should keep the params).

I think it's an important work and you have to start on solid basis, be sure that all services prepared by @jahow are well enough designed to cover all needs on how to init a map and how to feed it. I would add a permalinkService as well to manage this, there are some exemples in ngeo.

Reply:
After a couple of more hours of testing, the viewerService does not work as far as all the app is not using it I think. For example, getOpenedTool is watch, but as far as activeTools may be set when you click on a tool, the openTool object does not change. So you can have the first call to openTool(layers) works properly, switch to another tool using the UI eg. print, then openTool(layers) will not be watched. It would require to be unset openedTool.name = '' or used everywhere.

We probably need to improve the map viewer but we would not make for 3.4.0 I think. BTW, it would be nice to have this PR part of it I think - what do you/other/all think ?